bluetoothServicesComponent/modem_info_service.c

```diff
@@ -15,59 +15,67 @@
 #include "modem_info_service.h"
 #include "org.bluez.GattCharacteristic1.h"
 #include "org.bluez.GattService1.h"
+#include "org.bluez.GattDescriptor1.h"
 
 #define MODEM_INFO_FSN_CHARACTERISTIC_UUID "2A25"
-#define MODEM_INFO_IMEI_CHARACTERISTIC_UUID "2A27"
+#define MODEM_INFO_IMEI_CHARACTERISTIC_UUID "865d"
+#define CHARACTERISTIC_PRESENTATION_FORMAT_UUID "2904"
 
-struct FSNContext {
+static gboolean handle_read_fsn_value(
+    BluezGattCharacteristic1 *interface,
+    GDBusMethodInvocation *invocation,
+    GVariant *options,
+    gpointer user_data)
+{
     gchar fsn[32];
-    BluezGattCharacteristic1 *bt_characteristic;
-};
+    le_info_GetPlatformSerialNumber(fsn, 32);
+    g_print("%s called with FSN: %s\n", __func__, fsn);
+    GVariant *value = g_variant_new_bytestring((const gchar *)fsn);
+    g_variant_ref_sink(value);
+    bluez_gatt_characteristic1_set_value(interface, value);
+    bluez_gatt_characteristic1_complete_read_value(interface, invocation, value);
+    g_variant_unref(value);
 
-struct IMEIContext {
-    gchar imei[32];
-    BluezGattCharacteristic1 *bt_characteristic;
-};
+    return TRUE;
+}
 
-static gboolean handle_read_fsn_value(
+static gboolean handle_read_imei_value(
     BluezGattCharacteristic1 *interface,
     GDBusMethodInvocation *invocation,
     GVariant *options,
     gpointer user_data)
 {
-    struct FSNContext *ctx = user_data;
-    memset(ctx->fsn, 0, 32);
-    le_info_GetPlatformSerialNumber(ctx->fsn, 32);
-    g_print("%s called with FSN: %s\n", __func__, ctx->fsn);
-    GVariant *value = g_variant_new_bytestring((const gchar *)ctx->fsn);
+    gchar imei[32];
+    le_info_GetImei(imei, 32);
+    g_print("%s called with IMEI: %s\n", __func__, imei);
+    GVariant *value = g_variant_new_bytestring ((const gchar *)imei);
     g_variant_ref_sink(value);
     bluez_gatt_characteristic1_set_value(interface, value);
     bluez_gatt_characteristic1_complete_read_value(interface, invocation, value);
     g_variant_unref(value);
 
     return TRUE;
 }
-/* FSN */
 
-static gboolean handle_read_imei_value(
+static gboolean handle_read_cpf_value(
     BluezGattCharacteristic1 *interface,
     GDBusMethodInvocation *invocation,
     GVariant *options,
     gpointer user_data)
 {
-    struct IMEIContext *ctx = user_data;
-    // memset(ctx->imei, 0, 32);
-    le_info_GetImei(ctx->imei, 32);
-    g_print("%s called with IMEI: %s\n", __func__, ctx->imei);
-    GVariant *value = g_variant_new_bytestring ((const gchar *)ctx->imei);
+    guint8 custom_format[] = {25, 1, 0x00, 0x27, 1, 0x00, 0x00};
```

Comment:
You should leave a comment explaining the meaning of each of the values within the array since it won't be obvious to future readers of this code.